README.md
# PopWorldEditor

PopWorldEditor is a multipurpose tool used in creating maps for **Populous: The Beginning**.  

This repository has been made to improve upon ALACN's original work by fixing bugs, adding quality of life updates, or completely new features.
There are multiple forks of the original project out there, but this is the most up-to-date repository, although the commit history is missing due to this being uploaded as a new project.

### New Features
- **Network support over UDP**
- **8 players support (beta only)**
- **Copy, rotate and paste land with objects (CTRL + C, CTRL + V)**
- **Copying the entire map (CTRL + A)**
- **Importing maps (land) over an existing map**
- **Undo/redo for landscape and/or objects changes**
- **Many tools that automate tasks**
- **Map rotation (R)**
- **Shifting entire map to the left/right (SHIFT + PLUS/MINUS)**
- **New objects such as "Forbidden" or "Decorations" category**
- **Levl3 (beta) format support**
- **New Landscape brushes**
- **Reset selected markers with DELETE button**
- **Ability to quickly place markers (CTRL + LMB)**
- **Quick duplicate for objects (N key)**
- **General fixes**
- **Much more CPU usage friendly**
- **More QoL features**

<p align="center">
 <img width="800" height="600" src="https://i.imgur.com/XMQzAXF.gif"></img>
</p>


changelog.TXT
Version 1.9.0
Date: Aug 28 2021
Author: Toksisitee

* New Max. Alt brush setting. The value you enter is the max height (altitude) terrain can reach while terraforming.
* The World Editor will no longer crash (screen going black) when opening other full screen applications.
* Engine's X & Z coordinates are no longer being shown in the bottom left corner. People were mistakenly using these values in their scripts.
* When copying a marker position a comma will be added between the coords (e.g. "110, 100").
* The Main Menu bar cannot be closed any longer.
* New maps will now default to texture slot 1 instead of 0.
* Map Type & Objects Bank dialogs were merged together.


Version 1.8.0
Date: Jan 3 2020
Authors: Toksisitee & WhitePhantom

* Swap Tribe window received an overhaul and now supports all 8 tribes. (WhitePhantom)
* Cyan, Magenta, Black and Orange tribes were added to the Info window. (WhitePhantom)
* Cleaned up the arrangement of the Level Info window (WhitePhantom)
* CTRL + X command for cutting land pieces as presets.
* After opening the Brush window or entering Landscape Edit mode, you may use the arrow keys to move around.
* When entering Landscape Edit mode, the Brush window will automatically open.
* Renamed "AI Script" to "PopScript (Legacy)".
* Renamed "Script2" to "Lua Script".
* Updated About Window.


Version: 1.7.0
Date: Dec 3 2020
Author: Toksisitee

* Removed console window.
* You can now undo/redo object changes. (e.g. changing the owner, moving an object, etc.)
* Undo/redo buffers are now being cleared when opening/creating a new map.
* Canceling Land Preset mode (RMB) will no longer lower the land in the brush radius for a split second.
* Land Preset preview will be cleared upon changing from Landscape mode to something else (e.g to Objects or Markers mode).


Version: 1.6.0
Date: Nov 30 2020
Author: Toksisitee

* When copying land as a preset, you may now also copy objects. To enable/disable this, go to Options -> Brush -> Copy Objects.
* Show land height in the bottom left corner when in Landscape edit mode.
* Fixed a crash that would happen when trying to duplicate objects (via Object -> More -> Duplicate) without entering the amount.


Version: 1.5.0
Date: Oct 25 2020
Author: Toksisitee

* The map coordinates are displayed in the bottom left corner when in Landscape edit mode.
* Under the "Edit" menu, a new "Tools" menu item can be found. Some of these are NOT network synced.
  * (New) "Rotate Buildings": randomly rotate placed buildings.
  * (New) "Rotate Trees": randomly rotate placed trees.
  * (New) "Delete Duplicate Trees": delete stacked trees.
  * (New) "Delete Wildmen": delete all existing wildmen on the map.
  * (New) "Delete All Objects": delete all objects on the map.
  * (New) "Neutralizer": change the owner of all wildmen and scenery to neutral.
  * (New) "Levl3 to Levl2 Convert": change the owner of all objects to match the levl2 format. (e.g. cyan -> blue)
  * "Reset Markers": was moved under this category.
  * "Swap Tribes": was moved under this category.
  * "Macro": was moved under this category.


Version: 1.4.0
Date: Oct 16 2020
Author: Toksisitee & MrKosjaK

* Added sprites and objects previews for 8 player tribes (contribution by MrKosjaK).
* Added toggleable option that centers objects when saving the map under Options -> Misc -> Center Things (contribution by MrKosjaK).
* Added a permanent command console.
* Added a change history which allows the user to undo (CTRL + Z) and redo (CTRL + Y) land modifications.
* Added the ability of copying land pieces as presets by pressing CTRL + C. Anything in the brush radius will be copied.
* Added the ability to copy the entire map (land) by pressing CTRL + A.
* Added the ability to paste land presets by pressing CTRL + V (to enter Land Preset mode), then left click to place the land or right click to cancel.
* Added the ability of rotating the entire map (land) or land preset by pressing R. If the user is in Land Preset mode, it'll rotate the preset, otherwise the entire map.
* Added the ability of shifting the entire map (land) to the left/right by pressing SHFIT + PLUS/MINUS.


Version: 1.3.0
Date: Jun 26 2019
Author: Toksisitee

* Added Network support to the engine, this allows map editing with others over UDP.
* Reduced extremely high CPU usage from 20% ~ 30% down to 1%.
* Added a new object category -- "Decorations". 
    - It consists of 91 models, including previously inaccessible objects. They rotate like Scenery objects.
* Added a new object under the "Scenery" category.
    - "Forbidden" is an object that blocks the path of persons.
    - Persons can fly over it.
    - It cannot be destroyed, but it can be sinked.
* Added an "Highlight" button to the "Objects" menu that disables/enables highlighting of decoration objects.
* Fixed the "Shots" menu; now all objects under this category work.
* The "Objects" menu will now display the model ID & IslandNum of the selected object in the window title.
* Restore a marker's position by pressing the DELETE key while having a marker selected.
* Delete an object by selecting the object and pressing the DELETE key.
* Create a duplicate of the selected object by pressing the N key.
* Added 2 new landscape brushes; Raise & Lower.
* Increased max. brush size to 32.
* When you select a marker, you now have the option to copy the coordinates to clipboard with a button press.
* Added "Delete Objects" option, that will delete all objects on the map.
* While the "Markers" menu is open, you can hold CTRL and press the LMB to place the next  marker.
* When choosing to save or open a map, the default folder will be the directory of the World Editor.
* Added Level2 format types to "Save" & "Open" options. 
* Fixed Menu toolbar height.

network.h
#pragma once
#include "pop.h"

extern ENetAddress	  address;
extern ENetHost		  *server;
extern ENetEvent	    event;
extern ENetPacket	  *packet;
extern ENetHost		  *client;
extern ENetPeer			*peer;

#ifdef _DEBUG

#define _NET_DEBUG_CONNECTION			1
#define NET_CONNECT_IP					"localhost"
#define	NET_PORT						(1234)

#define	_NET_DEBUG_LAND					1

#endif

#define LAND_PACKETS_BUFFER				4096 // LAND_PACKETS / 2. 
#define NET_CONNECTION_RETRY_TIMER		3000
#define NET_SERVER_MAX_CONNECTIONS		1
#define NET_CLIENT_MAX_CONNECTIONS		1
#define NET_MAX_CHANNELS				1
#define NET_CLIENT						(0)
#define NET_SERVER						(1)
#define NET_ERROR						(2)
#define NET_SUCCESS						(3)
#define NET_DEBUG						(4)

#define SZ_NET_CLIENT					"[Client]"
#define SZ_NET_SERVER					"[Server]"
#define SZ_NET_ERROR					"[Error]"
#define SZ_NET_SUCCESS					"[Success]"
#define SZ_NET_DEBUG					"[Debug]"

#define PACKETTYPE_CREATE_OBJECT		(0)
#define PACKETTYPE_DELETE_OBJECT		(1)
#define PACKETTYPE_MOVE_OBJECT			(2)
#define PACKETTYPE_OWNER_OBJECT			(3)
#define PACKETTYPE_TYPE_OBJECT			(4)
#define	PACKETTYPE_MODEL_OBJECT			(5)
#define PACKETTYPE_ROTATE_OBJECT		(6)
#define PACKETTYPE_SWAP_TRIBE			(7)
#define PACKETTYPE_NEW_MAP_ASK			(8)
#define PACKETTYPE_NEW_MAP				(9)
#define PACKETTYPE_LAND_MODIFY			(10)
#define PACKETTYPE_UPDATE_VIEW			(11)
#define PACKETTYPE_AVAILABLE_DISC		(12)
#define PACKETTYPE_SPELL_RESTRICTIONS	(13)
#define PACKETTYPE_ALLIES				(14)
#define PACKETTYPE_NUM_PLAYERS			(15)
#define	PACKETTYPE_FOG_OF_WAR			(16)
#define PACKETTYPE_OMNIPRESENCE			(17)
#define	PACKETTYPE_NO_GUEST				(18)
#define	PACKETTYPE_NO_REINCARNATE_TIME	(19)
#define PACKETTYPE_COMPUTER_PLAYER_AI	(20)
#define	PACKETTYPE_OBJECT_BANK			(21)
#define PACKETTYPE_MAP_TYPE				(22)
#define PACKETTYPE_CREATE_MARKER		(23)
#define PACKETTYPE_MOVE_MARKER			(24)
#define PACKETTYPE_DELETE_MARKER		(25)
#define PACKETTYPE_TRIGGER_OPTIONS		(26)
#define PACKETTYPE_DISCOVERY_OPTIONS	(27)
#define PACKETTYPEMSG_DISCOVERY_MANA	(28)
#define PACKETTYPESER_SYNC_MAP			(29)
#define PACKETTYPE_SYNC_MAP_ASK			(30)
#define	PACKETTYPE_SYNC_MAP				(31)
#define PACKETTYPE_SYNC_MAP_UPDATE_USER	(32)
#define PACKETTYPE_SYNC_OBJECTS			(33)
#define PACKETTYPEMSG_SYNC_MISC			(34)
#define PACKETTYPE_TRIGGER_LINK			(35)
#define PACKETTYPE_DELETE_ALL_OBJECTS   (36)
#define PACKETTYPE_DELETE_ALL_MARKERS   (37)

struct Packet
{
	Packet()
	{
	
		
	}

	void del() 
	{
		delete this;
	}

	WORD wType;
	WORD wData[13];
};

struct LandPacket
{
	LandPacket()
	{
		clear();
	}

	WORD wType;
	WORD wData[LAND_PACKETS_BUFFER];
	void clear() { memset(this->wData, 0, sizeof(this->wData)); }
};


class Network 
{
public:
	Network(): m_bIsHost { false }, m_bENetInitialized { false }, 
		m_bConnected { false }, m_nConnections { 0 }, m_bNewMap { false }, 
		m_LandPackets { 0 }, m_nRecvWaitCount { 0 }, m_szIP { 0 }, m_nPort { 0 }
    {

    }
	void InitNetwork(bool bHosting);
	void SendPacket(WORD* p, size_t nSize);
	void SendPacket(const char* szMsg);
	void SendPacket(Packet *p, size_t nSize = sizeof(struct Packet));
	void SendPacket(LandPacket *p, size_t nSize = sizeof(struct LandPacket));
	bool IsHost(void) { return this->m_bIsHost; }
	bool IsInitialized(void) { return this->m_bENetInitialized; }
	bool IsConnected(void) { return this->m_bConnected; }
	BYTE GetConnections(void) { return this->m_nConnections; }
	void Connected(void) { this->m_nConnections++; this->m_bConnected = this->m_nConnections; }
	void Disconnected(void) { this->m_nConnections--; this->m_bConnected = this->m_nConnections; }
	bool GetNewMap() { return this->m_bNewMap; }
	void SetNewMap(bool accept) { this->m_bNewMap = accept; }
	void OnSyncMapAsk();
	int m_LandPackets;
private:
	template<class... Args>
	void Log(BYTE nLogType, Args... args);

	void InitConsole(void);
	bool ProcessHost(void);
	bool ProcessClient(void);
	bool m_bIsHost;
	bool m_bENetInitialized;
	bool m_bConnected;
	BYTE m_nConnections;
	std::thread m_thProcess;
	bool m_bNewMap;
	BYTE m_nRecvWaitCount;
	BYTE m_nSerializePacketType;
	char m_szIP[255];
	int m_nPort;

	void OnPacket(ENetEvent &event, bool bMsg = false, SBYTE nPacketType = -1);
	THING* OnCreateObject(BYTE nType, BYTE nModel, BYTE nOwner, WORD x, WORD z);
	void OnObjectDelete(WORD wThingIdx);
	void OnObjectMove(WORD wThingIdx, WORD x, WORD z, bool bIsLandbridge);
	void OnObjectOwner(WORD wThingIdx, BYTE nOwner);
	void OnObjectType(WORD wThingIdx, BYTE nType, BYTE nModel);
	void OnObjectModel(WORD wThingIdx, BYTE nModel, BYTE nIslandNum);
	void OnSwapTribe(BYTE t1, BYTE t2, BYTE t3, BYTE t4, bool f);
	void OnObjectRotate(WORD wThingIdx, BYTE nType, WORD wAngle);
	void OnEngineSetGroundHeight(int x, WORD h, int nNewPacket);
	void OnEngineUpdateView(void);
	void OnNewMapAsk(void);
	void OnNewMap(void);
	void OnAvailableDiscoveries(WPARAM wParam);
	void OnSpellRestrictions(WPARAM wParam);
	void OnAllies(WPARAM wParam);
	void OnNumPlayers(int nPlayers);
	void OnFogOfWar(bool bFog);
	void OnOmnipresence(bool bOmnipresence);
	void OnNoGuest(bool bGuestSpells);
	void OnNoReincarnationTime(bool bReincarnationTime);
	void OnComputerPlayerIndex(BYTE nTribe, int nScriptIndex);
	void OnObjectBank(BYTE nBank);
	void OnMapType(BYTE nBank);
	void OnDeleteMarker(int nMarker);
	void OnMarkerMove(int nMarker, WORD x, WORD z);
	void OnCreateMarker(int nMarker, WORD x, WORD z);
	void OnTriggerOptions(WORD wThingIdx, BYTE nType, BYTE nCellRadius, SBYTE nOccurences, WORD wTriggerCount, WORD wPrayTime, WORD wInactiveTime, bool bStartInactive, bool bCreateOwned);
	void OnDiscoveryOptions(WORD wThingIdx, BYTE nAvailability, BYTE nType, BYTE nModel);
	void OnDiscoveryMana(WORD wThingIdx, int nManaAmount);
	void OnPrepareUpdateMap(BYTE nRecvWaitCount, BYTE nSerializePacketType );
	void OnSyncMap(WORD* wLandData);
	void OnSyncUpdateUser(void);
	void OnSyncObjects(BYTE nType, BYTE nModel, BYTE nOwner, WORD x, WORD z, WORD wData5, WORD wData6, WORD wData7, WORD wData8, WORD wData9, WORD wData10, WORD wData11, SBYTE wData12);
	void OnSyncMisc(ULONG dwAvailableSpells, ULONG dwBuildingsAvailable, UBYTE nAllies0, UBYTE nAllies1, UBYTE nAllies2, UBYTE nAllies3, UBYTE nNumPlayers, UBYTE nLevelFlags, UBYTE nCP0, UBYTE nCP1, UBYTE nCP2, UBYTE nObjBank, UBYTE nMapBank, DWORD dwSpellsRestrictions);
	void OnTriggerLink(bool link, WORD wSourceIdx, WORD wTargetIdx, BYTE nSlot);
	void OnRemoveAllObjects(void);
	void OnRemoveAllMarkers(void);
}; 

extern Network net;